library/core/src/num/int_sqrt.rs

```diff
@@ -0,0 +1,190 @@
+
+                let numerator = ((r_prime as $FullBitsT) << QUARTER_BITS) | (lo >> QUARTER_BITS);
+                let denominator = (s_prime as $FullBitsT) << 1;
+
+                let q = numerator / denominator;
+                let u = numerator % denominator;
+
+                let mut s = (s_prime << QUARTER_BITS) as $FullBitsT + q;
+                if ((u << QUARTER_BITS) | (lo & ((1 << QUARTER_BITS) - 1))) < q * q {
+                    s -= 1;
+                }
+                s >> (precondition_shift >> 1)
+            };
+
+            result
+        }
+
+        const fn $intermediate_fn(mut n: $FullBitsT) -> ($FullBitsT, $FullBitsT) {
+            // Performs a Karatsuba square root.
+            // https://web.archive.org/web/20230511212802/https://inria.hal.science/inria-00072854v1/file/RR-3805.pdf
+
+            const HALF_BITS: u32 = <$FullBitsT>::BITS >> 1;
+            const QUARTER_BITS: u32 = <$FullBitsT>::BITS >> 2;
+
+            let leading_zeros = n.leading_zeros();
+            let result = if leading_zeros >= HALF_BITS {
+                let (s, r) = $intermediate_half_fn(n as $HalfBitsT);
+                (s as $FullBitsT, r as $FullBitsT)
+            } else {
+                // Either the most-significant bit or its neighbor must be a one, so we shift left to make that happen.
+                let precondition_shift = leading_zeros & (HALF_BITS - 2);
+                n <<= precondition_shift;
+
+                let hi = (n >> HALF_BITS) as $HalfBitsT;
+                let lo = n & (<$HalfBitsT>::MAX as $FullBitsT);
+
+                let (s_prime, r_prime) = $intermediate_half_fn(hi);
+
+                let numerator = ((r_prime as $FullBitsT) << QUARTER_BITS) | (lo >> QUARTER_BITS);
+                let denominator = (s_prime as $FullBitsT) << 1;
+
+                let q = numerator / denominator;
+                let u = numerator % denominator;
+
+                let mut s = (s_prime << QUARTER_BITS) as $FullBitsT + q;
+                let (mut r, overflow) =
+                    ((u << QUARTER_BITS) | (lo & ((1 << QUARTER_BITS) - 1))).overflowing_sub(q * q);
+                if overflow {
+                    r = r.wrapping_add((s << 1) - 1);
+                    s -= 1;
+                }
+                (s >> (precondition_shift >> 1), r >> (precondition_shift >> 1))
+            };
+
+            result
+        }
+    };
+}
+
+karatsuba_isqrt!(u16, u16, intermediate_u16, u8, u8, intermediate_u8);
+karatsuba_isqrt!(u32, u32, intermediate_u32, u16, u16, intermediate_u16);
+karatsuba_isqrt!(u64, u64, intermediate_u64, u32, u32, intermediate_u32);
+karatsuba_isqrt!(u128, u128, _intermediate_u128, u64, u64, intermediate_u64);
+
+#[cfg(target_pointer_width = "16")]
+#[inline(always)]
+pub const fn usize(n: usize) -> usize {
+    u16(n as u16) as usize
+}
+
+#[cfg(target_pointer_width = "32")]
+#[inline(always)]
+pub const fn usize(n: usize) -> usize {
+    u32(n as u32) as usize
+}
```

Comment:
Could you add doc comments for these helper functions? Better not to shadow the types too since it can get confusing, even if that means another field for the macro (names like `usize_` are fine, but a short prefix that hints what it does is better).

Also I think the functions `u8` and `intermediate_u8` can probably be grouped with the other functions

Reply:
> Better not to shadow the types too since it can get confusing, even if that means another field for the macro (names like `usize_` are fine, but a short prefix that hints what it does is better).

I'm unclear what you mean here. For the name, do you mean `usize_isqrt` or something like that? Do you mean another field for the `int_macros.rs` and `uint_macros.rs` macros?

I was essentially copying the naming scheme of `library/core/num/int_log10.rs`, which is done that way because the `int` and `uint` macros both have an `ident` with the type name (or an equivalent) in it, and that can be used to call a function with that `ident`'s name.

> Also I think the functions `u8` and `intermediate_u8` can probably be grouped with the other functions

Unfortunately, I think it can't be. The Karatsuba square root algorithm is recursive (the `u64` version will call the `u32` version and so forth) and the `u8` base case needs to be handled specially since there's no smaller type's implementation to call.

Reply:
> I'm unclear what you mean here. For the name, do you mean `usize_isqrt` or something like that? Do you mean another field for the `int_macros.rs` and `uint_macros.rs` macros?
> 
> I was essentially copying the naming scheme of `library/core/num/int_log10.rs`, which is done that way because the `int_impl!` and `uint_impl!` macros both have an `ident` with the type name (or an equivalent) in it, and that can be used to call a function with that `ident`'s name.

Ah I didn't realize this was being consistent, especially since the macro already takes separate `$fn` and `$half_fn` arguments. Thanks for clarifying.

`u64(i)` just looks like a cast rather than a square root call, but makes sense why it is that way - too bad we don't have `${concat(...)}` yet.

> Do you mean grouping the macro-created functions together and separately grouping the other functions together? I'll do that.

This one, simply rearranging the lines so the small functions are all together and the bigger macro is separate. (Feel free to ignore this one, just makes it a bit easier to read imo).

Reply:
OK, I've put the standalone functions on top, then the signed function macro and its functions, then all the macros for the unsigned functions and their functions.